R/dentist_qc.R

```diff
@@ -70,11 +73,180 @@ dentist <- function(zScore, LDmat, nSample,
+  }
+  return(merged_results)
+}
+
+dentist_detect_outliers <- function(sum_stat, LDmat, nSample,
+                                    window_size = 2000000, pValueThreshold = 5e-8, propSVD = 0.4, gcControl = FALSE,
+                                    nIter = 10, gPvalueThreshold = 0.05, ncpus = 1, seed = 999, correct_chen_et_al_bug = TRUE) {
+  # detect for column names and order by pos
+  if (!any(tolower(c("pos", "position")) %in% tolower(colnames(sum_stat))) ||
+    !any(tolower(c("z", "zscore")) %in% tolower(colnames(sum_stat)))) {
+    stop("Input sum_stat is missing either 'pos'/'position' or 'z'/'zscore' column.")
+  }
+  sum_stat <- sum_stat %>% arrange(pos)
+  ### FIXME: formalize this logic somewhere
+  ### if correct_chen_et_al_bug = FALSE, then there can never be a single window
+  # 1. if window_size is invalid (<=0 or >=range), and
+  # 1.1. if correct_chen_et_al_bug==TRUE, run dentist single window, so just window (0-4122)
+  # 1.2. if correct_chen_et_al_bug==FALSE, run in the dentist original way, so window (0-4122) and (742-4122)
+  # 2. if window_size is valid, and
+  # 2.1 if correct_chen_et_al_bug==TRUE, divide the window but don't add additional window if the first window covers all, so just (0-4122)
+  # 2.2 if correct_chen_et_al_bug==FALSE, divide the window in the dentist original way, so window (0-4122) and (742-4122)
+  if (window_size <= 0 | ((window_size >= max(sum_stat$pos) - min(sum_stat$pos) | is.na(window_size)) & (correct_chen_et_al_bug == TRUE))) {
+    print("single window!")
+    imputed_result <- dentist_impute_single_window(
+      sum_stat$z, LDmat, nSample,
+      pValueThreshold, propSVD, gcControl,
+      nIter, gPvalueThreshold, ncpus, seed, correct_chen_et_al_bug
+    )
+  } else {
+    # divide windows
+    window_divided_res <- divide_into_windows(sum_stat$pos, window_size = window_size, correct_chen_et_al_bug = TRUE)
+    print(window_divided_res)
+    # compute dentist result for each window
+    imputed_result_by_window <- list()
+    for (k in 1:nrow(window_divided_res)) {
+      zScore_k <- sum_stat$z[window_divided_res$windowStartIdx[k]:window_divided_res$windowEndIdx[k]]
+      LDmat_k <- LDmat[
+        window_divided_res$windowStartIdx[k]:window_divided_res$windowEndIdx[k],
+        window_divided_res$windowStartIdx[k]:window_divided_res$windowEndIdx[k]
+      ]
+      imputed_result_by_window[[k]] <- dentist_impute_single_window(
+        zScore_k, LDmat_k, nSample,
+        pValueThreshold, propSVD, gcControl,
+        nIter, gPvalueThreshold, ncpus, seed, correct_chen_et_al_bug
+      )
+    }
+    # merge imputed result and generate a final imputed_result (similar to imputed_result above)
+    imputed_result <- merge_windows(imputed_result_by_window, window_divided_res)
+  }
+  # detect outlier
+  lambda_original <- 1
+  imputed_result <- imputed_result %>%
+    mutate(
+      stat = calculate_stat(original_z, imputed_z, rsq),
```

Comment:
Please rename this -- from `stat` to `outlier_stat`

Reply:
renamed.